### PR TITLE
📉 Use `num_logits_to_keep` to reduce memory usage in GRPO

### PR DESCRIPTION
**Pull request by qgallouedec**
Don't compute the logits of the prompt, to save memory

```python
import torch
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B")
prompt_completion_ids = torch.randint(0, 100, (8, 256))
prompt_length = 100
completion_length = 156

# Previous approach

def get_per_token_logps_prev(model, input_ids):
    logits = model(input_ids).logits  # (B, L, V)
    logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred
    input_ids = input_ids[:, 1:]  # (B, L-1), exclude the first input ID since we don't have logits for it
    # Compute the log probabilities for the input tokens. Use a loop to reduce memory peak.
    per_token_logps = []
    for logits_row, input_ids_row in zip(logits, input_ids):
        log_probs = logits_row.log_softmax(dim=-1)
        token_log_prob = torch.gather(log_probs, dim=1, index=input_ids_row.unsqueeze(1)).squeeze(1)
        per_token_logps.append(token_log_prob)
    return torch.stack(per_token_logps)

per_token_logps_prev = get_per_token_logps_prev(model, prompt_completion_ids)
# Get rid of the prompt (-1 because of the shift done in get_per_token_logps)
per_token_logps_prev = per_token_logps_prev[:, prompt_length - 1 :]


# New approach

def get_per_token_logps(model, input_ids, num_logits_to_keep):
    # We add 1 to `num_logits_to_keep` because the last logits of the completion is later excluded
    logits = model(input_ids, num_logits_to_keep=num_logits_to_keep + 1).logits  # (B, L, V)
    logits = logits[:, :-1, :]  # (B, L-1, V), exclude the last logit: it corresponds to the next token pred

    # Compute the log probabilities for the input tokens. Use a loop to reduce memory peak.
    per_token_logps = []
    for logits_row, input_ids_row in zip(logits, input_ids[:, -num_logits_to_keep:]):
        log_probs = logits_row.log_softmax(dim=-1)
        token_log_prob = torch.gather(log_probs, dim=1, index=input_ids_row.unsqueeze(1)).squeeze(1)
        per_token_logps.append(token_log_prob)
    return torch.stack(per_token_logps)


num_logits_to_keep = completion_length  # we only need to compute the logits for the completion tokens
per_token_logps = get_per_token_logps(model, prompt_completion_ids, num_logits_to_keep)

assert torch.all(per_token_logps_prev  == per_token_logps)
```

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -427,29 +427,28 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
         completion_ids = prompt_completion_ids[:, prompt_length:]
 
         # Get the per-token log probabilities for the completions for the model and the reference model
-        def get_per_token_logps(model, input_ids):
-            logits = model(input_ids).logits  # (B, L, V)
+        def get_per_token_logps(model, input_ids, num_logits_to_keep):
+            # We add 1 to `num_logits_to_keep` because the last logits of the completion is later excluded
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:431]:
```suggestion
            # We add 1 to `num_logits_to_keep` because the last logits of the sequence is later excluded
```

**Comment by qgallouedec** [comment]
```python
from datasets import load_dataset
from trl import GRPOConfig, GRPOTrainer

dataset = load_dataset("trl-lib/tldr", split="train")

# Make everything deterministic
import torch

torch.manual_seed(0)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False


# Dummy reward function: the closer the completion is to 20 characters, the higher the reward
def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]


training_args = GRPOConfig(
    output_dir="Qwen2.5-0.5B-GRPO-main",
    logging_steps=2,
    gradient_accumulation_steps=1,
    per_device_train_batch_size=3,
    max_steps=20,
)
trainer = GRPOTrainer(
    model="Qwen/Qwen2.5-0.5B-Instruct",
    reward_funcs=reward_len,
    args=training_args,
    train_dataset=dataset,
)
trainer.train()
```

OOM with main

W B Chart 29_01_2025, 16_58_38

**Review by kashif — APPROVED** [review]

**Review by plaguss — APPROVED** [review]
